CS483/CS483/Kartaclysm/Components/ComponentTrack.cpp

```diff
@@ -0,0 +1,234 @@
+		{
+			if (strcmp(strTrackPieceId, m_vTrackPieces[i]->GetGUID().c_str()) == 0)
+			{
+				return i;
+			}
+		}
+
+		return -1;
+	}
+
+	int ComponentTrack::GetNextTrackPieceIndex(int p_iCurrentTrackPieceIndex)
+	{
+		int iNextTrackPieceIndex = p_iCurrentTrackPieceIndex + 1;
+		return iNextTrackPieceIndex >= m_vTrackPieces.size() ? 0 : iNextTrackPieceIndex;
+	}
+
+	int ComponentTrack::GetRacerIndex(const std::string& p_strRacerId)
+	{
+		const char* strRacerId = p_strRacerId.c_str();
+		for (unsigned int i = 0; i < m_vRacers.size(); ++i)
+		{
+			if (strcmp(strRacerId, m_vRacers[i].id.c_str()) == 0)
+			{
+				return i;
+			}
+		}
+
+		return -1;
+	}
+
+	void ComponentTrack::UpdateRacerPositions()
+	{
+		// TODO: optimize
+		// Matt: This is currently O(n^2).  While n will always be small, and it likely won't be a problem,
+		//			this can still be improved.  Something to pick up on a rainy day.
+		for (int i = 1; i < m_vRacers.size(); ++i)
+		{
+			RacerData racerA = m_vRacers[i];
+			bool bRacerPositionUpdated = false;
+			for (int j = i - 1; j >= 0; --j)
+			{
+				RacerData racerB = m_vRacers[j];
+				if (IsAhead(racerA, racerB))
+				{
+					m_vRacers[i] = racerB;
+					m_vRacers[j] = racerA;
+					bRacerPositionUpdated = true;
+					TriggerRacerPositionUpdateEvent(racerB.id);
+				}
+			}
+
+			if (bRacerPositionUpdated)
+			{
+				TriggerRacerPositionUpdateEvent(racerA.id);
+			}
+		}
+	}
+
+	// TODO: clean this up
+	// Matt: I don't really like the implementation of this function, but it will work for now
+	bool ComponentTrack::IsAhead(const RacerData& p_RacerA, const RacerData& p_RacerB)
+	{
+		// ahead by lap
+		if (p_RacerA.currentLap > p_RacerB.currentLap)
+		{
+			return true;
+		}
+		else if (p_RacerA.currentLap < p_RacerB.currentLap)
+		{
+			return false;
+		}
+		else
+		{
+			// ahead by track piece
+			if (p_RacerA.currentTrackPiece > p_RacerB.currentTrackPiece)
+			{
+				return true;
+			}
+			else if (p_RacerA.currentTrackPiece < p_RacerB.currentTrackPiece)
+			{
+				return false;
+			}
+			else
+			{
+				// ahead by distance
+				int iNextPieceIndex = GetNextTrackPieceIndex(p_RacerA.currentTrackPiece);
+
+				glm::vec3 vNextTrackPiecePosition = m_vTrackPieces[iNextPieceIndex]->GetTransform().GetTranslation();
+				glm::vec3 vRacerPosition = m_pGameObject->GetManager()->GetGameObject(p_RacerA.id)->GetTransform().GetTranslation();
+				glm::vec3 vOpponentPosition = m_pGameObject->GetManager()->GetGameObject(p_RacerB.id)->GetTransform().GetTranslation();
+
+				return glm::distance(vNextTrackPiecePosition, vRacerPosition) < glm::distance(vNextTrackPiecePosition, vOpponentPosition);
+			}
+		}
+	}
+
+	void ComponentTrack::TriggerRacerPositionUpdateEvent(const std::string& p_strRacerId)
+	{
+		HeatStroke::Event* pEvent = new HeatStroke::Event("RacerPositionUpdated");
+		pEvent->SetStringParameter("racerId", p_strRacerId);
+		HeatStroke::EventManager::Instance()->TriggerEvent(pEvent);
+	}
+
```

Comment:
I feel like this could attach more parameters like the position it was changed from/to. Point being that in general it doesn't cost anything performance-wise to have events be as verbose as possible for whatever the receiver may need.